dev-docs/yaml-design-guidance.md

```diff
@@ -0,0 +1,312 @@
+
+Make sure you:
+1. Mention the change in the changelog
+2. Update docs
+3. Update any test configuration
+4. Create a migration test as detailed in [migrations](apollo-router/src/configuration/migrations/README.md)
+5. In your migration description tell the users what they have to update.
+
+## Process
+It should be obvious to the user what they are configuring and how it will affect Router behaviour. It's tricky for us as developers to know when something isn't obvious to users as often we are too close to the domain.
+
+Complex configuration changes should be discussed with the team before starting the implementation, since they will drive the code's design. The process is as follows:
+1. In the github issue put the proposed config in.
+2. List any concerns.
+3. Notify the team that you are looking for request for comment.
+4. Ask users what they think.
+5. If you are an Apollo Router team member then schedule a meeting to discuss. (This is important, often design considerations will fall out of conversation)
+6. If it is not completely clear what the direction should be:
+7. Wait a few days, often people will have ideas later even if they didn't in the meeting.
+8. Make your changes.
+
+Note that these are not hard and fast rules, and if your config is really obviously correct then by all means make the change and be prepared to deal with comments at the review stage.
+
+## Design patterns
+
+Use the following as a rule of thumb, also look at existing config for inspiration.
+The most important goal is usability, so do break the rules if it makes sense, but it's worth bringing the discussion to the team in such circumstances.  
+
+1. [Avoid empty config](#avoid-empty-config).
+2. [Do use `#[serde(deny_unknown_fields)]`](#do-use-serdedeny_unknown_fields).
+3. [Don't use `#[serde(flatten)]`](#dont-use-serdeflatten).
+4. [Use consistent terminology](#use-consistent-terminology).
+5. [Don't use negative options](#dont-use-negative-options).
+6. [Document your configuration options](#document-your-configuration-options).
+7. [Plan for the future](#plan-for-the-future).
+
+### Avoid empty config
+
+In Rust you can use `Option` to say that config is optional, however this can give a bad experience if the type is complex and all fields are optional.
+
+#### GOOD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url // url is required
+}
+```
+```yaml
+export:
+    url: http://example.com
+```
+
+#### GOOD
+```rust
+enum ExportUrl {
+    Default,
+    Url(Url)
+}
+
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: ExportUrl // Url is required but user may specify `default`
+}
+```
+```yaml
+export:
+    url: default
+```
+
+#### GOOD
+In the case where you genuinely have no config or all sub-options have obvious defaults then use an `enabled: bool` flag.
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    enabled: bool,
+    url: Optional<Url> // url is optional
```

Comment:
Is it intended that you use `Optional<>` instead of `Option<>` ?
Also I'm wondering if in that case if it has a default value we should force people to not use optional field but use the field with a serde attribute `#[serde(default = "default_resource")]` What do you think ?

Reply:
It depends, `default_resource` is OK if the answer is static, not it if is something like an env variable as generating the schema will leak the value.
In this case it's probably OK. I'll update and then merge.